### PR TITLE
Improve DAOs

### PR DESCRIPTION
**Pull request by simonbrandhof**
Note that branch is badly named.
- add RuleDao to sonar-db. First implementation to replace sonar-server deprecated RuleDao
- use DatabaseUtils.executeLargeInputs() some DAOs

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/issue/IssueService.java [hunk 1]
```diff
@@ -289,7 +289,7 @@ public Issue getByKey(String key) {
   IssueDto getByKeyForUpdate(DbSession session, String key) {
     // Load from index to check permission : if the user has no permission to see the issue an exception will be generated
     Issue authorizedIssueIndex = getByKey(key);
-    return dbClient.issueDao().selectByKeyOrFail(session, authorizedIssueIndex.key());
+    return dbClient.issueDao().selectOrFailByKey(session, authorizedIssueIndex.key());
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/IssueService.java]: I LOVE that name method : D

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivatorContextFactory.java [hunk 2]
```diff
@@ -76,12 +76,12 @@ private RuleActivatorContext create(RuleKey ruleKey, DbSession session, RuleActi
   }
 
   private RuleDto initRule(RuleKey ruleKey, RuleActivatorContext context, DbSession dbSession) {
-    RuleDto rule = db.ruleDao().getNullableByKey(dbSession, ruleKey);
+    RuleDto rule = db.deprecatedRuleDao().getNullableByKey(dbSession, ruleKey);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivatorContextFactory.java]: rename by selectByKey ?

File: sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java [hunk 3]
```diff
@@ -78,23 +78,31 @@ public ActionPlanDto selectByKey(String key) {
...
     } finally {
       MyBatis.closeQuietly(session);
     }
   }
 
+  public List<ActionPlanDto> selectByKeys(DbSession dbSession, Collection<String> keys) {
+    return DatabaseUtils.executeLargeInputs(keys, new SelectByKeys(dbSession.getMapper(ActionPlanMapper.class)));
+  }
+
+  private static class SelectByKeys implements Function<List<String>, List<ActionPlanDto>> {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java]: name it KeysToActionPlanDtosFunction ?

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java [hunk 4]
```diff
@@ -89,18 +90,23 @@ public DefaultIssueComment selectCommentByKey(String commentKey) {
...
   }
 
-  List<IssueChangeDto> selectByIssuesAndType(DbSession session, Collection<String> issueKeys, String changeType) {
-    if (issueKeys.isEmpty()) {
-      return Collections.emptyList();
+  public List<IssueChangeDto> selectByTypeAndIssueKeys(DbSession session, Collection<String> issueKeys, String changeType) {
+    return DatabaseUtils.executeLargeInputs(issueKeys, new SelectByIssueKeys(session.getMapper(IssueChangeMapper.class), changeType));
+  }
+
+  private static class SelectByIssueKeys implements Function<List<String>, List<IssueChangeDto>> {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: name it KeysToIssueChangeDtosFunction ?

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java [hunk 5]
```diff
@@ -89,18 +90,23 @@ public DefaultIssueComment selectCommentByKey(String commentKey) {
     }
   }
 
-  List<IssueChangeDto> selectByIssuesAndType(DbSession session, Collection<String> issueKeys, String changeType) {
-    if (issueKeys.isEmpty()) {
-      return Collections.emptyList();
+  public List<IssueChangeDto> selectByTypeAndIssueKeys(DbSession session, Collection<String> issueKeys, String changeType) {
+    return DatabaseUtils.executeLargeInputs(issueKeys, new SelectByIssueKeys(session.getMapper(IssueChangeMapper.class), changeType));
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: I like the convention to have a mapper(session) method for each dao

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivatorContextFactory.java (hunk 2 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivatorContextFactory.java]: no, name can't be changed as it's "daoV2"

File: sonar-db/src/main/java/org/sonar/db/rule/RuleDao.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  /**
+   * Select rules by keys, whatever their status. Returns an empty list
+   * if the list of {@code keys} is empty, without any db round trip.
+   */
+  public List<RuleDto> selectByKeys(final DbSession session, List<RuleKey> keys) {
+    return executeLargeInputs(keys, new SelectByKeys(session.getMapper(RuleMapper.class)));
+  }
+
+  private static class SelectByKeys implements Function<List<RuleKey>, List<RuleDto>> {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleDao.java]: name it RuleKeysToRuleDtosFunction ?

File: sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java (hunk 3 above)
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java]: No, this function is specific to this method and isn't/won't be reused elsewhere.

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java (hunk 4 above)
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: No, this function is specific to this method and isn't/won't be reused elsewhere.

File: sonar-db/src/main/java/org/sonar/db/user/UserDao.java [hunk 7]
```diff
@@ -74,27 +77,34 @@ public UserDto selectActiveUserByLogin(DbSession session, String login) {
...
 
-  public List<UserDto> selectUsersByLogins(DbSession session, List<String> logins) {
-    List<UserDto> users = Lists.newArrayList();
-    if (!logins.isEmpty()) {
-      UserMapper mapper = session.getMapper(UserMapper.class);
-      List<List<String>> partitions = Lists.partition(logins, 1000);
-      for (List<String> partition : partitions) {
-        users.addAll(mapper.selectUsersByLogins(partition));
-      }
+  private static class SelectByLogins implements Function<List<String>, List<UserDto>> {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/user/UserDao.java]: name it LoginsToUserDtosFunction ?

File: sonar-db/src/test/java/org/sonar/db/rule/RuleDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Rule
+  public DbTester dbTester = DbTester.create(System2.INSTANCE);
+
+  RuleDao underTest = dbTester.getDbClient().ruleDao();
+
+  @Test
+  public void selectByKeys() {
```
Comment (teryk) [inline sonar-db/src/test/java/org/sonar/db/rule/RuleDaoTest.java]: add a test to test selectByKeys(key) ?

**Comment by teryk** [comment]
apart from the naming conventions, LGTM!

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java (hunk 5 above)
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: Fixed

File: sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java (hunk 3 above)
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/issue/ActionPlanDao.java]: I understand that, but still I think the KeysToActionPlanDtosFunction is clearer. I haven't suggested ActionPlanKeysToActionPlanDtosFunction.

**Pull request closed without merging**